jf_agent/git/github_gql_client.py

```diff
@@ -0,0 +1,466 @@
+        """
+        try:
+            return self._get_raw_result(query_body=query_body)['data']['organization']
+        except Exception as e:
+            log_and_print_request_error(
+                e, f'fetching source project {login}. ' f'Skipping...',
+            )
+            return None
+
+    def get_users(self, login: str) -> Generator[dict, None, None]:
+        query_body = f"""{{
+            organization(login: \"{login}\") {{
+                userQuery: membersWithRole(first: 100, after: %s) {{
+                    {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                    users: nodes {{
+                        {self.GITHUB_GQL_USER_FRAGMENT}
+                    }}
+                }}
+            }}
+        }}
+        """
+        for page in self._page_results(
+            query_body=query_body, path_to_page_info='data.organization.userQuery'
+        ):
+            for user in page['data']['organization']['userQuery']['users']:
+                yield user
+
+    def get_repos(
+        self, login: str, repo_filters: list[filter] = None
+    ) -> Generator[dict, None, None]:
+        query_body = f"""{{
+            organization(login: "{login}") {{
+                repoQuery: repositories(first: 50, after: %s) {{
+                    {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                    repos: nodes {{
+                        ... on Repository {{
+                            id: databaseId
+                            name
+                            fullName: nameWithOwner
+                            url
+                            isFork
+                            defaultBranch: defaultBranchRef {{
+                                    name
+                            }}
+                            # This should be broken out into a separate query if 
+                            # hasNextPage is True.
+                            branchQuery: refs(refPrefix:"refs/heads/", first: 50) {{
+                                {self.GITHUB_GQL_PAGE_INFO_BLOCK}
+                                branches: nodes {{
+                                    ... on Ref {{
+                                        name
+                                        target {{sha: oid}}
+                                    }}
+                                }}
+                            }}
+                        }}
+                    }}
+                }}
+            }}
+        }}
+        """
+        for page in self._page_results(
+            query_body=query_body, path_to_page_info='data.organization.repoQuery'
+        ):
+            for api_repo in page['data']['organization']['repoQuery']['repos']:
+                # Skip over excluded or ignore non-included
+                if not all(filt(api_repo) for filt in repo_filters):
+                    continue
+                else:
+                    # Potentially process more branches
+                    if api_repo['branchQuery']['pageInfo']['hasNextPage']:
```

Comment:
I can see this logic getting a bit confusing in the future. Would be better to separate out our processing of branches into a separate flow? In ideal state we'd be submitting branches to a queue to process when we learn about them. I don't think we should change anything now but starting a discussion.

Reply:
Yeah this code assumes that for most repos that we process there are actually not more 50 branches on it (this repo only has 18, for example). It is a little confusing, but I think it makes sense to do this way. Branches are super simple data objects (NormalizedBranch only has a `name` and `sha` field) and the `NormalizedBranch` class is only used as a subclass on `NormalizedRepository`. Intertwining them here makes sense to me, and it gives us a potential speed boost if we guess correctly that there less than 50 branches in a repo.